onnxruntime/test/util/compare_ortvalue.cc

```diff
@@ -453,9 +453,17 @@ std::pair<COMPARE_RESULT, std::string> VerifyValueInfo(const ONNX_NAMESPACE::Val
     // TODO: CXX API doesn't have IsTensorSequence() supported for Ort::Value
     // TODO: Repeat whatever we did for Tensor above in a loop ?
     return std::make_pair(COMPARE_RESULT::SUCCESS, "");
-  }
+  } else if (v.type().has_optional_type()) {
+    const auto& tp = v.type().optional_type().elem_type();
+
+    if (tp.has_tensor_type() && !o.IsTensor()) {
+      return std::make_pair(COMPARE_RESULT::TYPE_MISMATCH, "");
+    }
+
```

Comment:
If has_tensor_type() returns true we check the shape info. Is there a reason we don't do that here?

Reply:
Check shape info for an optional tensor that is not None